src/common/heartbeat/src/listener.cpp

```diff
@@ -1,26 +1,138 @@
+#include <iostream>
+#include <sstream>
+
 #include <rclcpp/rclcpp.hpp>
-#include <std_msgs/msg/bool.hpp>
+#include <builtin_interfaces/msg/time.hpp>
+#include <std_msgs/msg/empty.hpp>
 
-#include <iostream>
+#include "common_interfaces/msg/e_stop.hpp"
+#include "common_interfaces/srv/handshake.hpp"
+#include "common_interfaces/srv/heartbeat_control.hpp"
+#include "heartbeat/date.h"
+
+// TODO: ros params instead of this
+#define PERIOD 3000 // how long should the time between pings be? (negotiated during handshake)
+#define TIMEOUT_MS 3500 // how long is too long between pings?
+#define WATCHDOG_MS 100 // how often should the watchdog check time between pings?
 
 class Listener : public rclcpp::Node {
 public:
-    Listener() : Node("heartbeat_lsitener") {
+    Listener() : Node("heartbeat_listener") {
         using namespace std::chrono_literals;
-        heartbeat_subscription = this->create_subscription<std_msgs::msg::Bool>("heartbeat", 10, std::bind(&Listener::heartbeat, this, std::placeholders::_1));
+        // create handshake client and control service
+        handshake_client = this->create_client<common_interfaces::srv::Handshake>("heartbeat_handshake");
+        control_service = this->create_service<common_interfaces::srv::HeartbeatControl>("heartbeat_control", std::bind(&Listener::heartbeat_control, this, std::placeholders::_1, std::placeholders::_2));
+
+        // create subscribers to handle incoming messages
+        heartbeat_sub = this->create_subscription<builtin_interfaces::msg::Time>("heartbeat", 10, std::bind(&Listener::ensure_alive, this, std::placeholders::_1));
+
+        // create estop publisher
+        estop_pub = this->create_publisher<common_interfaces::msg::EStop>("estop", 10);
+
+        // request producer's start
+        handshake();
+
+        // create timeout watchdog
+        watchdog_timer = this->create_wall_timer(std::chrono::milliseconds(WATCHDOG_MS), std::bind(&Listener::watchdog, this));
     }
 private:
-    void heartbeat(std_msgs::msg::Bool::ConstSharedPtr msg) {
-        std::cout << "Beat" << std::endl;
+    void restart(const std_msgs::msg::Empty::SharedPtr msg) {
+        (void) msg;
+        // retry handshake
+        handshake();
+    }
+
+    void handshake() {
+        using namespace date;
+        auto req = std::make_shared<common_interfaces::srv::Handshake::Request>();
+
+        while(!handshake_client->wait_for_service(std::chrono::milliseconds(100))) {
+            if (!rclcpp::ok()) {
+                RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Interrupted while waiting for the <%s> service. Exiting.", handshake_client->get_service_name());
+                return;
+            }
+            RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "<%s> Service not available, waiting again...", handshake_client->get_service_name());
```

Comment:
The RCLCPP_INFO macro has a "throttle_duration_sec" (I think is the exact name) that will drop some of these messages if they come in really fast. Since you're checking every 100ms, throttling to once every 1 or even every 5 seconds would help save the terminal

Reply:
You could also use RCLCPP_DEBUG... but then again it might be useful to know if the connection isn't working, so leaving it isn't a bad idea.

Reply:
Will change to RCLCPP_INFO_THROTTLE